### PR TITLE
[SPARK-21041][SQL] SparkSession.range should be consistent with SparkContext.range

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR fixes the inconsistency in `SparkSession.range`.

**BEFORE**
```scala
scala> spark.range(java.lang.Long.MAX_VALUE - 3, java.lang.Long.MIN_VALUE + 2, 1).collect
res2: Array[Long] = Array(9223372036854775804, 9223372036854775805, 9223372036854775806)
```

**AFTER**
```scala
scala> spark.range(java.lang.Long.MAX_VALUE - 3, java.lang.Long.MIN_VALUE + 2, 1).collect
res2: Array[Long] = Array()
```

## How was this patch tested?

Pass the Jenkins with newly added test cases.

### CODE REVIEW

**Comment by rednaxelafx** [comment]
The end result looks good to me. Thanks for fixing it!

Although, I'd prefer fixing the actual integer overflow handling in `RangeExec`'s codegen, too. Even though your fix will handle the problematic case for sure, it still feels a bit awkward to rely on an optimization for correctness.

**Comment by dongjoon-hyun** [comment]
Thanks. For me, this optimizer does more than the fixing. You can make another PR for the underlying codegen if you want.

**Comment by kiszk** [comment]
Interesting discussion.
@dongjoon-hyun Would it be possible to share among us your thought on why this optimizer does more than fixing the codegen?

**Comment by dongjoon-hyun** [comment]
Sure!

First of all, this handles another invalid case, Range(0,0). 

Second, although we fix the generated codes in the reported in SPARK-21041, we still generate **No-Op** code.

Third, this PR removes the usuless No-Op codegen and Janino compilation and so on. This also will cowork with the other optimizer like PripagateEmptyRelation. 

Finally, we can simplify the underlying logic later by preventing this kind of corner cases as eariler as possible.

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan and @gatorsmile .
Could you review this PR?

**Comment by kiszk** [comment]
@dongjoon-hyun thank you. IMHO, the codegen can handle the first point. However, the second and third points make sense. The codegen cannot achieve.
Finally, it seems to be design decision about what is the Catalyst optimizer role. Only optimization, or optimization and correctness. 
I would like to hear opinions of @cloud-fan and @gatorsmile, too.

**Comment by wzhfy** [comment]
Should we just invalidate such cases? Like we do for step cannot be 0.

**Comment by dongjoon-hyun** [comment]
That's also a good another candidate. In this PR, I just didn't want to surprise the users with the behavior's changes.

**Comment by gatorsmile** [comment]
@dongjoon-hyun Conceptually, we should not fix the correctness by adding optimization rules. 

- Fix the issues when [building the `Range`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala#L484-L492). 
- Also need to add the extra checking logics [here] (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala#L502) to block all the invalid cases.

**Comment by dongjoon-hyun** [comment]
I want to make another PR for the fixing bug. Is it okay for you all?

**Comment by dongjoon-hyun** [comment]
Hmm. Ah, sorry. I'll update this PR since the optimizer is not valid anymore.

**Comment by dongjoon-hyun** [comment]
@gatorsmile Are you sure to raise exception on empty range, Range(0, 0), really?

**Comment by wzhfy** [comment]
Yea, let's update this pr so that previous discussions stay in the same thread.

**Comment by dongjoon-hyun** [comment]
Yep, I'm updating here, @wzhfy .
I will close the SPARK-21044 as Invalid after this is merged.

**Comment by wzhfy** [comment]
@dongjoon-hyun Based on our definition of range, since the default step is 1, I think range(0, 0) is invalid. It's similar to the case when step is 0: range(1, 1, 0).

**Comment by dongjoon-hyun** [comment]
The problem is Spark RDD supports that in that way until now. We cannot raise exceptions on Dataset operations inconsistently.
```
scala> sc.range(0,0).collect
res1: Array[Long] = Array()

scala> sc.range(1,10,-1).collect
res2: Array[Long] = Array()
```

**Comment by rxin** [comment]
Sorry it doesn't make sense to do this. Range is used primarily for testing, and it doesn't make sense to have an optimizer rule that removes it. If there is a correctness issue in it, we should fix that.

**Comment by dongjoon-hyun** [comment]
@rxin . Yep. The scope is changed like that.

**Comment by dongjoon-hyun** [comment]
Thank you for review again, @rednaxelafx , @kiszk , @wzhfy , @gatorsmile , @rxin .
Now, the corner cases are replaced at the factory and the behavior are consistent.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -482,11 +482,17 @@ case class Sort(
 
 /** Factory for constructing new `Range` nodes. */
 object Range {
-  def apply(start: Long, end: Long, step: Long, numSlices: Option[Int]): Range = {
+  def apply(start: Long, end: Long, step: Long, numSlices: Option[Int])
+      : LeafNode with MultiInstanceRelation = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: this signature looks weird...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 2]
```diff
@@ -482,11 +482,17 @@ case class Sort(
...
     val output = StructType(StructField("id", LongType, nullable = false) :: Nil).toAttributes
-    new Range(start, end, step, numSlices, output)
+    if (start == end || (start < end ^ 0 < step)) {
+      LocalRelation(output)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Can we have something like `Range.empty` to still return a `Range`?

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Thank you for review, @cloud-fan . In that case, which spec can we use for that empty `Range` which is returned from `Range.empty`? For the following cases, it seems that we need to remove one of the following precondition again.
```
  require(step != 0, s"step ($step) cannot be 0")
  require(start != end, s"start ($step) cannot be equal to end ($end)")
  require(start < end ^ step < 0, s"the sign of step ($step) is invalid for range ($start, $end)")
```

**Comment by dongjoon-hyun** [comment]
@cloud-fan . For the signature, I used `LogicalPlan` instead.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 3]
```diff
@@ -500,6 +504,8 @@ case class Range(
   extends LeafNode with MultiInstanceRelation {
 
   require(step != 0, s"step ($step) cannot be 0")
+  require(start != end, s"start ($start) cannot be equal to end ($end)")
+  require(start < end ^ step < 0, s"the sign of step ($step) is invalid for range ($start, $end)")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: If we allow to return an empty relation/empty range in Range.apply, why we won't allow it too here? Although it's no-op, but looks like it's legal query?

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: This precheck statements are added by review comments. This explains the assumptions used in the implementations.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 2 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: I'd agree that an empty `Range` seems a more reasonable choose than a `LocalRelation` to return for those cases.

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: What is the values of parameters, start, end, step?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 3 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Is it? I saw the comments we should generate no-op code if the query is no-op? And you also said we cannot raise exceptions on Dataset operations inconsistently?

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: We already prevent them in the above factory method.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: I see. You assume we should only use factory `Range` to construct the nodes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 2 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: It looks a bit weird to see a `LocalRelation` node when constructing a `Range` node.

If possible, I'd let the empty `Range` keeps its wrong parameters as is. So the logical plan can be consistent with input query. And only turning it to empty relation when planning the query.

Of course this involves few change in Range's planning. So I'm ok with current solution too.

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Actually, at the first commit, I handled this invalid 'Range' in a new optimizer, 'RemoveInvalidRange'. That would be a similar approach.

I will wait more comment on this~ Thank you for review.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Hmm, I think doing it in optimization sounds not ok for me too. It's not an optimization actually. Anyway, let's wait for other comments. Maybe there's other good ways to do.

Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: I like empty `Range` keeps its wrong parameters as is, too.

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Then do you mean removing the 'require' statements back? For the invalid parameters, 'require' raises exceptions now.

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: @kiszk . What is the difference in your `empty` Range and invalid Range? It looks to me the same in your suggestion.

**Comment by cloud-fan** [comment]
I think `Range` is mostly for testing, and we don't need to worry about performance here, especially for empty range.

How about we just improve the `Range` to produce empty result for invalid parameters?

**Comment by dongjoon-hyun** [comment]
Current codegen implementation already generate empty result for the invalid case in general. This is about a trivial corner case. Is this corner case worth of making the codegen implement more complex?

**Comment by cloud-fan** [comment]
is it very hard? Seems we can just output 0-partition RDD for invalid parameters in `RangeExec`.

**Comment by dongjoon-hyun** [comment]
I see, @cloud-fan. I'll update like that.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 4]
```diff
@@ -347,8 +347,12 @@ case class RangeExec(range: org.apache.spark.sql.catalyst.plans.logical.Range)
   }
 
   override def inputRDDs(): Seq[RDD[InternalRow]] = {
-    sqlContext.sparkContext.parallelize(0 until numSlices, numSlices)
-      .map(i => InternalRow(i)) :: Nil
+    val rdd = if (start == end || (start < end ^ 0 < step)) {
+      new EmptyRDD[InternalRow](sqlContext.sparkContext)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Thank you for this idea.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameRangeSuite.scala [hunk 5]
```diff
@@ -191,6 +191,17 @@ class DataFrameRangeSuite extends QueryTest with SharedSQLContext with Eventuall
...
     }
   }
+
+  test("SPARK-21041 SparkSession.range()'s behavior is inconsistent with SparkContext.range()") {
+    val start = java.lang.Long.MAX_VALUE - 3
+    val end = java.lang.Long.MIN_VALUE + 2
+    Seq("false", "true").foreach { value =>
+      withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> value) {
+        assert(spark.sparkContext.range(start, end, 1).collect.length == 0)
+        assert(spark.range(start, end, 1).collect.length == 0)
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameRangeSuite.scala]: Shall we also test the case `start == end`?

Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameRangeSuite.scala]: Sure.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/2.2

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan !
Also, thank you for all reviewers.

**Pull request closed without merging**